Add api.structured; switch structured kernels to use const Tensor& everywhere

ezyang opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #51590 Support at::cpu on non-structured kernels
* #51585 Split anonymous and namespaced definitions in RegisterDispatchKey
* #51583 Factor out structured generation into its own subclass.
* #51508 Split out RegisterDispatchKey to its own file
* #51500 Use Literal to model targets.
* #51499 Add support for generating faithful at::cpu signatures
* **#51490 Add api.structured; switch structured kernels to use const Tensor& everywhere**
* #51477 Relax type signature for tools.codegen.api.translate

Mutable Tensor ref is a source of endless confusion for kernel writers;
if we're going to make everyone rewrite their kernels, might as well
also get rid of mutable Tensor& while we're at it.

This is a refactor-then-small-update double whammy.  The refactor
is to separate tools.codegen.api.structured from api.native for
describing the type signatures of structured kernels (previously,
I was naughtily reusing native for this purpose--now I need it to
behave differently as Tensor).  This started off as a copy paste, but
since there are not that many structured kernels so far I could delete
all of the legacy logic from native that didn't make sense (without
having to go out and fix all the use sites all at once).

One more small addition was teaching translate to convert Tensor& to const Tensor&.

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

Differential Revision: [D26182413](https://our.internmc.facebook.com/intern/diff/D26182413)

ezyang committed:
Switch structured kernels to use const Tensor& everywhere

Mutable Tensor ref is a source of endless confusion for kernel writers;
if we're going to make everyone rewrite their kernels, might as well
also get rid of mutable Tensor& while we're at it.

This is a refactor-then-small-update double whammy.  The refactor
is to separate tools.codegen.api.structured from api.native for
describing the type signatures of structured kernels (previously,
I was naughtily reusing native for this purpose--now I need it to
behave differently as Tensor).  This started off as a copy paste, but
since there are not that many structured kernels so far I could delete
all of the legacy logic from native that didn't make sense (without
having to go out and fix all the use sites all at once).

One more small addition was teaching translate to convert Tensor& to const Tensor&.

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

[ghstack-poisoned]

ezyang committed:
Update on "Switch structured kernels to use const Tensor& everywhere"


Mutable Tensor ref is a source of endless confusion for kernel writers;
if we're going to make everyone rewrite their kernels, might as well
also get rid of mutable Tensor& while we're at it.

This is a refactor-then-small-update double whammy.  The refactor
is to separate tools.codegen.api.structured from api.native for
describing the type signatures of structured kernels (previously,
I was naughtily reusing native for this purpose--now I need it to
behave differently as Tensor).  This started off as a copy paste, but
since there are not that many structured kernels so far I could delete
all of the legacy logic from native that didn't make sense (without
having to go out and fix all the use sites all at once).

One more small addition was teaching translate to convert Tensor& to const Tensor&.

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

Differential Revision: [D26182413](https://our.internmc.facebook.com/intern/diff/D26182413)

[ghstack-poisoned]

ezyang committed:
Update on "Switch structured kernels to use const Tensor& everywhere"


Mutable Tensor ref is a source of endless confusion for kernel writers;
if we're going to make everyone rewrite their kernels, might as well
also get rid of mutable Tensor& while we're at it.

This is a refactor-then-small-update double whammy.  The refactor
is to separate tools.codegen.api.structured from api.native for
describing the type signatures of structured kernels (previously,
I was naughtily reusing native for this purpose--now I need it to
behave differently as Tensor).  This started off as a copy paste, but
since there are not that many structured kernels so far I could delete
all of the legacy logic from native that didn't make sense (without
having to go out and fix all the use sites all at once).

One more small addition was teaching translate to convert Tensor& to const Tensor&.

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

Differential Revision: [D26182413](https://our.internmc.facebook.com/intern/diff/D26182413)

[ghstack-poisoned]

ezyang committed:
Update on "Switch structured kernels to use const Tensor& everywhere"


Mutable Tensor ref is a source of endless confusion for kernel writers;
if we're going to make everyone rewrite their kernels, might as well
also get rid of mutable Tensor& while we're at it.

This is a refactor-then-small-update double whammy.  The refactor
is to separate tools.codegen.api.structured from api.native for
describing the type signatures of structured kernels (previously,
I was naughtily reusing native for this purpose--now I need it to
behave differently as Tensor).  This started off as a copy paste, but
since there are not that many structured kernels so far I could delete
all of the legacy logic from native that didn't make sense (without
having to go out and fix all the use sites all at once).

One more small addition was teaching translate to convert Tensor& to const Tensor&.

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

Differential Revision: [D26182413](https://our.internmc.facebook.com/intern/diff/D26182413)

[ghstack-poisoned]

bhosmer reviewed (approved):
Nice! A couple factoring queries/comments, minor stuff. I'd definitely change the PR title to reflect its significance though - something like "Add structured API" :P

ezyang commented on tools/codegen/api/structured.py:
```diff
@@ -0,0 +1,94 @@
+    name = str(func.name.name)
+    if func.name.overload_name:
+        name += f'_{func.name.overload_name}'
+    return name
```
Actually this function is dead and I can just delete it.

ezyang commented on tools/codegen/api/structured.py:
```diff
@@ -0,0 +1,94 @@
+            )
+        # TODO: delete these special cases; see tools.codegen.api.cpp--these
+        # must be changed in tandem, but there are problems; see
+        # https://github.com/pytorch/pytorch/pull/51485
```
I don't think utility is necessarily the right way to fix this. In an ideal world, we would canonicalize all C++ types (e.g., expand type aliases, resolve namespaces) before doing conversions so it would be self evident that `IntArrayRef` is the same thing as `ArrayRef<int64_t>`. With this canonicalization, it wouldn't be necessary to have the special cases at all (ergo no utility). But for unfortunate reasons (people directly matching on the C++ type; see the pull request) this can't be done right now, and I don't really want to block this patch set on it. But I kind of want this to still be potentially painful to update so that if it is painful again in the future someone will get around to actually fixing this.

bhosmer replied:
I guess... my worry is that having two similar but not identical piles of conversions in different parts of the codebase will just look entropic (if it's noticed at all) and invite fixing things by hacking, e.g. a new conversion needs to be added but no structured kernels use it yet so don't bother adding it there, that kind of thing. Not a hill to die on though.

ezyang replied:
I agree, it absolutely is entropic. I am mostly suggesting that if we are worried about the entropy, it is better to invest time in fixing https://github.com/pytorch/pytorch/pull/51485

ezyang committed:
Update on "Switch structured kernels to use const Tensor& everywhere"


Mutable Tensor ref is a source of endless confusion for kernel writers;
if we're going to make everyone rewrite their kernels, might as well
also get rid of mutable Tensor& while we're at it.

This is a refactor-then-small-update double whammy.  The refactor
is to separate tools.codegen.api.structured from api.native for
describing the type signatures of structured kernels (previously,
I was naughtily reusing native for this purpose--now I need it to
behave differently as Tensor).  This started off as a copy paste, but
since there are not that many structured kernels so far I could delete
all of the legacy logic from native that didn't make sense (without
having to go out and fix all the use sites all at once).

One more small addition was teaching translate to convert Tensor& to const Tensor&.

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

Differential Revision: [D26182413](https://our.internmc.facebook.com/intern/diff/D26182413)

[ghstack-poisoned]
